rust/tw_scale/src/macros.rs

```diff
@@ -0,0 +1,346 @@
+use super::*;
+
+macro_rules! tuple_impl {
+    ($(,)?) => {};
+    ($first:tt, $($rest:tt),* $(,)?) => {
+        tuple_impl!($($rest),*,);
+        tuple_impl!(@make_impl $first $($rest)*);
+    };
+    (@make_impl $($t:tt)+) => {
+        #[allow(non_snake_case, unused_parens, unconditional_recursion)] // the compiler seems confused here
+        impl <$($t),+> ToScale for ($($t),+,) where $($t: ToScale),+ {
+            fn to_scale_into(&self, out: &mut Vec<u8>) {
+                let ($($t),*) = self;
+                $(
+                    $t.to_scale_into(out);
+                )*
+            }
+        }
+    };
+}
+
+//tuple_impl!(T0, T1, T2, T3, T4, T5, T6, T7);
+tuple_impl!(T0, T1, T2);
+
+#[macro_export]
+macro_rules! replace_ident {
+    ($_t:tt $sub:ident) => {
+        $sub
+    };
+}
+
+#[macro_export]
+macro_rules! replace_expr {
+    ($_t:tt $sub:expr) => {
+        $sub
+    };
+}
+
+#[macro_export]
+macro_rules! impl_struct_scale {
+  // "New Type" struct
+  (
+    $(#[$struct_meta:meta])*
+    pub struct $struct_name:ident (
+      $(#[$struct_field_meta:meta])*
+      $struct_field_vis:vis $struct_field_ty:ty
+    );
+  ) => {
+    $(#[$struct_meta])*
+    pub struct $struct_name(
+      $(#[$struct_field_meta])*
+      $struct_field_vis $struct_field_ty
+    );
+
+    impl $crate::ToScale for $struct_name {
+      fn to_scale_into(&self, out: &mut Vec<u8>) {
+        self.0.to_scale_into(out);
+      }
+    }
+  };
+  // Normal named fields struct.
+  ($(#[$struct_meta:meta])*
+  pub struct $struct_name:ident {
+    $(
+      $(#[$struct_field_meta:meta])*
+      $struct_field_vis:vis $struct_field_name:ident : $struct_field_ty:ty
+    ),+ $(,)?
+  }) => {
+    $(#[$struct_meta])*
+    pub struct $struct_name {
+      $(
+        $(#[$struct_field_meta])*
+        $struct_field_vis $struct_field_name : $struct_field_ty
+      ),+
+    }
+
+    impl $crate::ToScale for $struct_name {
+      fn to_scale_into(&self, out: &mut Vec<u8>) {
+        $(self.$struct_field_name.to_scale_into(out);)+
+      }
+    }
+  }
+}
+
+#[macro_export]
+macro_rules! impl_enum_scale {
```

Comment:
The reason for these macros is to avoid encoding errors from manually implementing `ToScale` for structs/enums.  The generated code will be the same as what would need to be manually written.

I was also thinking of supporting SCALE decoding.  Which would be needed for transaction decoding.

Reply:
I have simplified the `impl_enum_scale` macro by dropping support for automatic variant index.

Reply:
Thank you for the changes!